packages/services/api/src/modules/organization/providers/organization-access-tokens.ts

```diff
@@ -315,31 +567,533 @@ export class OrganizationAccessTokens {
+    await this.session.assertPerformAction({
+      organizationId: project.orgId,
+      params: { organizationId: project.orgId, projectId: project.id },
+      action: 'projectAccessToken:modify',
+    });
+
+    const accessToken = await this.getById(id);
+
+    if (!accessToken || accessToken?.projectId !== project.id) {
+      return null;
+    }
+
+    return accessToken;
+  }
+
+  async getAvailablePermissionsGroupsForProject(project: Project): Promise<Array<PermissionGroup>> {
+    await this.session.assertPerformAction({
+      organizationId: project.orgId,
+      params: { organizationId: project.orgId, projectId: project.id },
+      action: 'projectAccessToken:modify',
+    });
+
+    const organization = await this.storage.getOrganization({ organizationId: project.orgId });
+    const groups = await this.getAvailablePermissionGroupsForOrganization(organization);
+
+    return groups
+      .map(group => ({
+        ...group,
+        permissions: group.permissions.filter(permission =>
+          validProjectResourceLevels.has(getPermissionGroup(permission.id)),
+        ),
+      }))
+      .filter(group => group.permissions.length !== 0);
+  }
+
+  private createFeatureFlagPermissionFilter(organization: Organization) {
+    const isAppDeplymentsEnabled = organization.featureFlags.appDeployments;
+    const isOTELTracingEnabled = organization.featureFlags.otelTracing;
+
+    return (id: Permission) =>
+      (!isAppDeplymentsEnabled && id.startsWith('appDeployment:')) ||
+      (!isOTELTracingEnabled && id.startsWith('traces:'))
+        ? false
+        : true;
+  }
+
+  async getAvailablePermissionGroupsForOrganization(
+    organization: Organization,
+  ): Promise<Array<PermissionGroup>> {
+    const filter = this.createFeatureFlagPermissionFilter(organization);
+
+    return OrganizationAccessTokensPermissions.permissionGroups
+      .map(group => ({
+        ...group,
+        permissions: group.permissions
+          .filter(permission => filter(permission.id))
+          .map(permission => ({
+            ...permission,
+            isAssignableByViewer: true,
+          })),
+      }))
+      .filter(group => group.permissions.length !== 0);
+  }
+
+  async getAvailablePermissionGroupsForMembership(membership: OrganizationMembership) {
+    const organization = await this.storage.getOrganization({
+      organizationId: membership.organizationId,
+    });
+    const groups = await this.getAvailablePermissionGroupsForOrganization(organization);
+    const memberPermissions = membership.assignedRole.role.allPermissions;
+
+    return groups
+      .map(group => ({
+        ...group,
+        permissions: group.permissions.map(permission => ({
+          ...permission,
+          isAssignableByViewer: memberPermissions.has(permission.id),
+        })),
+      }))
+      .filter(group => group.permissions.length !== 0);
+  }
+
+  async getPermissionsForAccessToken(accessToken: OrganizationAccessToken) {
```

Comment:
Resolves the permissions for an access token, takes into account all constraints such as for PATs